Support empty record declaration

objmagic opened the pull request:
This PR adds support of empty record declaration.

Support empty record declaration can
1. bring more consistency in the language
2. solve some practical issues. For example, [`ocaml-pb`](https://github.com/yallop/ocaml-pb) can generate proper code for empty protobuf message.

See [Mantis #7583](https://caml.inria.fr/mantis/view.php?id=7583) for more details.

Implementation in this PR might be inefficient during type-checking phase because we will search empty record declaration in environment in linear time complexity (instead of other `find-` methods in `env.ml` which will use `Path.t` to do search in a AVL tree). That said, another approach is to allow `{}` as user-defined constructor as we did in https://github.com/ocaml/ocaml/pull/234

objmagic committed:
Allow empty variants and records.

objmagic committed:
work.

objmagic committed:
save.

objmagic committed:
Support empty record type.

objmagic committed:
Allow empty record in pattern

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
Allow empty record in pattern.

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
Pretty-print, fix merge error.

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
Use `Atom(0)' instead.

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
Add test cases.

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
Minor fixes.

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
Add test for undeclared record type.

Signed-off-by: Runhang Li <marklrh@gmail.com>

objmagic committed:
no empty variants

objmagic committed:
Merge branch 'trunk' into empty-record-variants

objmagic committed:
fix merge error

objmagic committed:
fix some tests

objmagic committed:
unbox status

objmagic committed:
improve dsource

alainfrisch commented on bytecomp/translcore.ml:
```diff
@@ -1284,7 +1284,11 @@ and transl_record loc env fields repres opt_init_expr =
         | Record_regular -> Lconst(Const_block(0, cl))
         | Record_inlined tag -> Lconst(Const_block(tag, cl))
-        | Record_unboxed _ -> Lconst(match cl with [v] -> v | _ -> assert false)
+        | Record_unboxed _ -> Lconst(
```
What's the point of "unboxing" an empty record type, since the representation ends up being the same as the regular representation?

alainfrisch commented on parsing/pprintast.ml:
```diff
@@ -1397,8 +1398,11 @@ and type_declaration ctxt f x =
     | Ptype_variant xs ->
-        pp f "%t%t@\n%a" intro priv
-          (list ~sep:"@\n" constructor_declaration) xs
+        if xs = [] then
```
This is about empty variants, not empty records.

alainfrisch commented on typing/oprint.ml:
```diff
@@ -592,9 +592,13 @@ and print_out_type_decl kwd ppf td =
-      fprintf ppf " =%a@;<1 2>%a"
-        print_private td.otype_private
-        (print_list print_out_constr (fun ppf -> fprintf ppf "@ | ")) constrs
+      if constrs = [] then
```
Also for empty variants, not records.

alainfrisch commented on typing/typecore.ml:
```diff
@@ -2952,7 +2963,21 @@ and type_expect_ ?in_function ?(recarg=Rejected) env sexp ty_expected =
-              None -> newvar (), None
+            | None ->
+                if lid_sexp_list = [] then begin
+                  match Env.find_empty_record env with
```
This part could be shared with the similar addition type type_pat.

alainfrisch commented on parsing/parser.mly:
```diff
@@ -1707,7 +1707,8 @@ expr_comma_list:
 ;
 record_expr:
-    simple_expr WITH lbl_expr_list              { (Some $1, $3) }
+                                                { (None, []) }
```
Empty rules usually have `/* empty */` in parser.mly.

objmagic commented:
@alainfrisch There are some changes about empty variants that I forgot to clean up. I will remove then address your comments next morning when I get up.

objmagic committed:
Write empty in empty rule, remove variants code

objmagic committed:
Refactor typecore

murmour commented on parsing/parsetree.mli:
```diff
@@ -407,9 +407,7 @@ and type_declaration =
 and type_kind =
   | Ptype_abstract
   | Ptype_variant of constructor_declaration list
-        (* Invariant: non-empty list *)
```
The comment related to `Ptype_variant` shouldn't be removed.

objmagic committed:
add back invariant comment.

objmagic committed:
Add change entry.

objmagic commented:
@murmour @alainfrisch Thanks for your review. I have addressed your comments.

I would like to have some comments on typing part, especially if we should switch to allow empty record as user-defined constructor because current implementation could be quite slow.

xavierleroy commented:
At the last developers's meeting, there was consensus to support empty sum types, but empty record types were controversial.  Have we really decided to support empty record types?  Could we get a decision before spending more time reviewing and polishing this GPR?

objmagic commented:
@xavierleroy 

Hi, Xavier. I understand empty record is controversial but I'd like to open a PR just for some initial discussion. As for empty sum type, I noticed some cases my implementation does not handle. I am going to fix the problem soon.

garrigue commented:
Ok, then I suppose that we should have a discussion.

Basically, I see at least 2 ways to handle empty records without adding anything to the language:

* Have your code generator emit `()` rather than `{}`. Doesn't seem too difficult.
  We could even make `{}` an alias for `()`, but some are going to complain that this is a waste of syntactic resources...

* If you want the thing to be nominal, make it a singleton variant type: `type t = Empty_t`.
  This even gives you a way to use the constructor without ambiguity.

Honestly I think that the first solution is sufficient: I cannot really imagine why you would want to make your empty record nominally distinct, as anyway it has no semantical contents.

Regularity is nice, when it is really regular. Adding empty records as records requires modifications to the way we handle records that make me think that they are not regular records.

yallop commented:
The following code is already valid OCaml (and introduces isomorphic/nominally-distinct types):
```ocaml
    type s = []
    type t = []
```
so it seems reasonable to allow the essentially equivalent code:
```ocaml
    type s = {}
    type t = {}
```

But since there are plenty of workarounds and a clear lack of enthusiasm it doesn't seem worth spending much more time on this.

objmagic commented:
close this PR as it is deemed controversial
